📖  update general/product documentation (#1135)
operator-framework/operator-controller: docs/index.md

* A namespace-based multi-tenancy model in lieu of namespace-scoped CRDs
* A packaging model in which catalogs of extensions, usually containing the entire version history of each extension, are made available to clusters for cluster users to browse and select from

Since its initial release, OLM has helped catalyse the growth of Operators throughout the Kubernetes ecosystem. [OperatorHub.io](https://operatorhub.io/)
is a popular destination for discovering Operators, and boasts over 300 packages from many different vendors.

# Why are we building OLM v1?

OLM (v0) has been in production for over 5 years, and the community to leverage this experience and question the initial
goals and assumptions of the project. OLM v1 is a complete redesign and rewrite of OLM taking into account this accumulated experience. 
Compared to its predecessor, amongst other things, OLM v1 aims to provide:
* A simpler API surface and mental model
* Less opinionated automation and greater flexibility
* Support for Kubernetes applications beyond only Operators
* Security by default
* Helm Chart support

> * GitOps support

Author: Yeah, let's push that effort to a different PR. It should mostly be rejiggering order, I'd say.

# What Operator Lifecycle Manager (OLM)?

Operator Lifecycle Manager (OLM) is an open-source [CNCF](https://www.cncf.io/) project with the mission to manage the 

> lifecycle of cluster extensions centrally and declaratively on Kubernetes clusters. Its is to make installing, 

Reviewer: ```suggestion
lifecycle of cluster extensions centrally and declaratively on Kubernetes clusters. Its purpose is to make installing, 
```
Or similar?

running, and updating functional extensions to the cluster easy, safe, and reproducible for cluster administrators and PaaS administrators, 
throughout the lifecycle of the underlying cluster.

> OLM was focused on a particular type of cluster extension: [operators](https://operatorhub.io/what-is-an-operator#:~:text=is%20an%20Operator-,What%20is%20an%20Operator%20after%20all%3F,or%20automation%20software%20like%20Ansible.). 
> An operator is a Kubernetes application composed of one or more controllers and one, or more, API extensions, 
> or [CustomResourceDefinitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)) (CRDs).

Reviewer: ```suggestion
OLM was focused on a particular type of cluster extension: [Operators](https://operatorhub.io/what-is-an-operator#:~:text=is%20an%20Operator-,What%20is%20an%20Operator%20after%20all%3F,or%20automation%20software%20like%20Ansible.). 
An Operator is a Kubernetes application composed of one or more controllers and one, or more, API extensions, 
or [CustomResourceDefinitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) (CRDs).
```

OLM helped define lifecycles for these extensions in which the extensions:

> * get installed, potentially causing other dependency extensions to be installed as well

Reviewer 2: I was under the impression that we are not doing any dependency resolution as part of OLM 1.0.

Author: I don't think it's our highest priority, but,  I also don't think it's been completely ruled out ^^ (I'm sure PM really wants it XD)

* can be customized through configuration at runtime

> * get upgraded to newer version/s following upgrade paths defined by extension developers
> * and finally, get decommissioned and removed.

Reviewer: ```suggestion
* get upgraded to newer version/s following upgrade paths defined by extension developers, and finally,
* get decommissioned and removed.
```
Might move this to the previous bullet
```suggestion
* get upgraded to newer version/s following upgrade paths defined by extension developers
* get decommissioned and removed
```
or just drop.

The first iteration of OLM, termed v0, included several concepts and features targeting the stability, security, and supportability of the life-cycled applications, for instance:

> * A dependency model that enabled cluster extensions to focus on their primary purpose by delegating out of scope behavior to dependencies.

Reviewer 2: right, we are acknowledging that in V1 we are doing dependency resolution.  IMO we can just remove it from the first section.

Author: I've condensed the previous section to something a little more general and left this section as is.

> * A dependency model that enabled cluster extensions to focus on their primary purpose by delegating out of scope behavior to dependencies.
> * A constraint model that allowed cluster extension developers to define support limitations such as conflicting extensions, and minimum kubernetes versions
> * A namespace-based multi-tenancy model in lieu of namespace-scoped CRDs.  
> * A packaging model in which catalogs of extensions, usually containing the entire version history of each extension, are made available to clusters for cluster users to browse and select from.

Reviewer: ```suggestion
* A dependency model that enabled cluster extensions to focus on their primary purpose by delegating out of scope behavior to dependencies
* A constraint model that allowed cluster extension developers to define support limitations such as conflicting extensions, and minimum kubernetes versions
* A namespace-based multi-tenancy model in lieu of namespace-scoped CRDs
* A packaging model in which catalogs of extensions, usually containing the entire version history of each extension, are made available to clusters for cluster users to browse and select from
```

> Since its initial release, OLM has helped catalyse the growth of operators throughout the Kubernetes ecosystem. [OperatorHub.io](https://operatorhub.io/)
> is a popular destination for discovering operators, and boasts over 300 packages from a many different vendors.

Reviewer: ```suggestion
Since its initial release, OLM has helped catalyse the growth of Operators throughout the Kubernetes ecosystem. [OperatorHub.io](https://operatorhub.io/)
is a popular destination for discovering Operators, and boasts over 300 packages from many different vendors.
```

# Why are we building OLM v1?

> OLM (v0) has been in production for over 5 years, and the community to leverage this experience and question the initial

Reviewer: ```suggestion
OLM v0 has been in production for over 5 years, and the community to leverage this experience and question the initial
```
to match the instances of v1 (which don't have parentheses)

> goals and assumptions of the project. OLM v1 is a complete redesign and rewrite of OLM taking into account this accumulated experience. 

Reviewer: I am having trouble grok'ing the first sentence, but I also might be reading it wrong.

Compared to its predecessor, amongst other things, OLM v1 aims to provide:
* A simpler API surface and mental model
* Less opinionated automation and greater flexibility

> * Support for Kubernetes applications beyond only operators

Reviewer: ```suggestion
* Support for Kubernetes applications beyond only Operators
```

* Security by default

> * Helm support

Reviewer 2: Wondering if this should be `Helm` or `Helm Chart` support.

* GitOps support

> For an in-depth look at OLM v1, please see the [OLM v1 Overview](olm_overview.md), and the [Roadmap](olmv1_roadmap.md).

Reviewer: ```suggestion
For an in-depth look at OLM v1, please see the [OLM v1 Overview](olm_overview.md) and the [Roadmap](olmv1_roadmap.md).
```

# The OLM community

> In this next iteration of OLM, the community has also taken care to make it as contributed friendly as possible, and welcomes new contributors. 

Reviewer: ```suggestion
In this next iteration of OLM, the community has also taken care to make it as contributor-friendly as possible, and welcomes new contributors. 
```
?

The project is tracked in a [GitHub project](https://github.com/orgs/operator-framework/projects/8/), 

> which provides a great entrypoint to quickly find something interesting to work on and contribute.

Reviewer: ```suggestion
which provides a great entry point to quickly find something interesting to work on and contribute.
```

You can reach out to the OLM community for feedbacks/discussions/contributions in the following channels:

>   * Kubernetes slack channel: [#olm-dev](https://kubernetes.slack.com/messages/olm-dev)
>   * [Operator Framework on Google Groups](https://groups.google.com/forum/#!forum/operator-framework)
>   * Weekly in Person Working Group Meeting: [olm-wg](https://github.com/operator-framework/community#operator-lifecycle-manager-working-group)

Reviewer: ```suggestion
  * Kubernetes Slack channel: [#olm-dev](https://kubernetes.slack.com/messages/olm-dev)
  * [Operator Framework on Google Groups](https://groups.google.com/forum/#!forum/operator-framework)
  * Weekly in-person Working Group meeting: [olm-wg](https://github.com/operator-framework/community#operator-lifecycle-manager-working-group)
```
